### ISSUE TITLE
Event loops & mutlithreading

### ISSUE THREAD

**Issue by zerodefect** [alanxz/SimpleAmqpClient#161]
Not sure this is an issue per se. Nonetheless, maybe someone else may find my question helpful.

As mentioned in the Readme, the SimpleAmqpClient is a wrapper around the [rabbitmq-c](https://github.com/alanxz/rabbitmq-c) library.  At this point in time, I'm looking to write a (relatively simple) producer - no message consumption required just yet.

One of the talking points listed in the C-api is the limitation (can I call it that? :) ) in using multiple threads.  Can I assume that the C++ variation inherits the same limitations? From here on, I assume this to be the case.

All being considered, I am proposing to use a C++11 std::thread that maintains all the state pertaining to the AMQP client (so everything is allocated within the thread).  The thread will have shared access to a custom thread-safe queue to consume requests and push binary data into the pertinent channels/queues.

One of the concerns with my proposal is that I would look to block the thread while waiting for data to be available to the thread-safe queue.  Is this dangerous? I ask because having searched around, I've seen mentions of event loops and the like. 

Any suggestions/tips would be much appreciated.  Once I receive an answer I may update the docs.

**Comment by alanxz** [comment]
> One of the talking points listed in the C-api is the limitation (can I call it that? :) ) in using multiple threads. Can I assume that the C++ variation inherits the same limitations? From here on, I assume this to be the case.

That is correct an `amqp_connection_state_t` is not thread-safe.  Access from multiple threads must be synchronized.  The same applies to the roughly equivalent `AmqpClient::Channel` class.

> One of the concerns with my proposal is that I would look to block the thread while waiting for data to be available to the thread-safe queue. Is this dangerous? I ask because having searched around, I've seen mentions of event loops and the like.

SimpleAmqpClient does not have an internal event-loop, so blocking outside of SimpleAmqpClient is fine.  

The talk about event loops centers around dealing with asynchronous events like having callbacks when messages are received, or implementing support for heartbeats, neither of which are implemented in SimpleAmqpClient.

**Comment by zerodefect** [comment]
Thanks for your prompt response.

> That is correct an amqp_connection_state_t is not thread-safe. Access from multiple threads must be synchronized. The same applies to the roughly equivalent AmqpClient::Channel class.

Your first inline comment raises some intrigue.  I had understood that that the same thread of execution must be used at all times.  Thinking about it, this probably only makes sense if one is using an event loop?

Following on from this, is it also safe too use an internal thread pool and produce messages from different threads in a synchronized fashion?  I would be looking to use a boost asio strand to perform the synchronization across different threads.

**Comment by alanxz** [comment]
> Your first inline comment raises some intrigue. I had understood that that the same thread of execution must be used at all times. Thinking about it, this probably only makes sense if one is using an event loop?

Not sure, I haven't thought deeply about when using an object would require always accessing it from the same thread. This isn't the case for SimpleAmqpClient.

> Following on from this, is it also safe too use an internal thread pool and produce messages from different threads in a synchronized fashion? I would be looking to use a boost asio strand to perform the synchronization across different threads.

As long as only 1 thread is calling into a given `Channel` object at a time, then yes that seems like it will be safe.

**Comment by capsocrates** [comment]
I have safely (as far as I can tell) used separate `Channel` objects on multiple threads concurrently with no issues.

**Comment by alanxz** [comment]
@capsocrates separate `Channel` objects can be used concurrently on different threads.  Note in the current version of SimpleAmqpClient, a channel object maps to a single connection to the broker.
